fe1-web/src/features/lao/reducer/ServerReducer.ts

```diff
@@ -0,0 +1,198 @@
+        byAddress: {},
+        allAddresses: [],
+        backendKeyByFrontendKey: {},
+      };
+
+      if (server.laoId in state.byLaoId) {
+        laoState = state.byLaoId[server.laoId];
+      }
+
+      if (server.address in laoState.byAddress) {
+        throw Error(
+          `${server.address} is already part of laoState.byAddress. Use updateServer() instead of addServer()`,
+        );
+      }
+
+      laoState.byAddress[server.address] = server;
+      laoState.allAddresses.push(server.address);
+
+      laoState.backendKeyByFrontendKey[server.frontendPublicKey] = server.serverPublicKey;
+      state.byLaoId[server.laoId] = laoState;
+    },
+
+    updateServer: (state: Draft<ServerReducerState>, action: PayloadAction<ServerState>) => {
+      const updatedServer = action.payload;
+
+      if (!(updatedServer.laoId in state.byLaoId)) {
+        return;
+      }
+
+      if (!(updatedServer.address in state.byLaoId[updatedServer.laoId].byAddress)) {
+        return;
+      }
+
+      // delete old frontend-backend key mapping entry
+      delete state.byLaoId[updatedServer.laoId].backendKeyByFrontendKey[
+        state.byLaoId[updatedServer.laoId].byAddress[updatedServer.address].frontendPublicKey
+      ];
+
+      // update byAddress entry
+      state.byLaoId[updatedServer.laoId].byAddress[updatedServer.address] = updatedServer;
+
+      // add new frontend-backend key mapping entry
+      state.byLaoId[updatedServer.laoId].backendKeyByFrontendKey[updatedServer.frontendPublicKey] =
+        updatedServer.serverPublicKey;
+    },
+
+    removeServer: (state, action: PayloadAction<{ laoId: string; address: string }>) => {
+      const { laoId, address: serverAddress } = action.payload;
+
+      if (!(laoId in state.byLaoId)) {
+        return;
+      }
+
+      if (serverAddress in state.byLaoId[laoId].byAddress) {
+        // delete the frontend-backend key mapping entry
+        delete state.byLaoId[laoId].backendKeyByFrontendKey[
+          state.byLaoId[laoId].byAddress[serverAddress].frontendPublicKey
+        ];
+
+        // cleanup the byAddress and allAddress entries
+        delete state.byLaoId[laoId].byAddress[serverAddress];
+        state.byLaoId[laoId].allAddresses = state.byLaoId[laoId].allAddresses.filter(
+          (address) => address !== serverAddress,
+        );
+      }
+    },
+
+    clearAllServers: (state) => {
+      state.byLaoId = {};
+    },
+  },
+});
+
+export const { addServer, clearAllServers, removeServer, updateServer } = serverSlice.actions;
+
+export const serverReduce = serverSlice.reducer;
+
+export default {
+  [SERVER_REDUCER_PATH]: serverSlice.reducer,
+};
+
+export const getServerState = (state: any): ServerReducerState => state[SERVER_REDUCER_PATH];
+
+/**
+ * A function to directly retrieve the public key from the redux store for a given lao id and server address
+ * @remark NOTE: This function does not memoize the result. If you need this, use makeServerSelector instead
```

Comment:
```suggestion
 * @remark This function does not memoize the result. If you need this, use makeServerSelector instead
```
No need to write both.